src/GridFS/WritableStream.php

```diff
@@ -262,7 +263,7 @@ private function fileCollectionInsert()
         $this->file['length'] = $this->length;
         $this->file['uploadDate'] = new UTCDateTime();
 
-        if (! $this->disableMD5) {
+        if (! $this->disableMD5 && $this->hashCtx) {
```

Comment:
Is this needed for psalm? The constructor always initializes `hashCtx` with `hash_init()` when `disableMD5` is `false`, so I'd expect this wouldn't be necessary unless psalm can't infer that `hash_init()` always returns a valid object.

That aside, should we use `!== null` or `instanceof` here instead of relying on a truthy cast?

Reply:
I don't think Psalm can be smart enough to detect that you never mutate `disableMD5è and to keep a relation between type refinements for that (it might be able to do that with a readonly property when it knows it cannot be mutated, but I don't expect it to ever be able to handle that for mutable properties)

Reply:
> any method call could reset it to null even with disableMD5 being false

Ah, and even though it doesn't happen in WritableStream the class isn't final so it theoretically _could_ happen in a child class. Similar explanation as some other questions -- I think I'm understanding the logic behind these warnings a bit better now. :+1: